flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AbstractStreamTaskNetworkInput.java

```diff
@@ -151,7 +152,11 @@ protected DataInputStatus processEvent(BufferOrEvent bufferOrEvent) {
         final AbstractEvent event = bufferOrEvent.getEvent();
         // TODO: with checkpointedInputGate.isFinished() we might not need to support any events on
         // this level.
-        if (event.getClass() == EndOfPartitionEvent.class) {
+        if (event.getClass() == EndOfData.class) {
+            if (checkpointedInputGate.hasReceivedEndOfData()) {
+                return DataInputStatus.END_OF_DATA;
+            }
```

Comment:
Why is this check on this level and not inside the input gate? If for nothing else, just for the sake of consistency with `EndOfPartitionEvent`? But it also sounds wrong. We are receiving `EndOfData` but not really.

Reply:
`EndOfPartitionEvent` is also handled in the `AbstractStreamTaskNetworkInput`, but for `EndOfPartitionEvent`, `CheckpointedInputGate` returns an empty buffer and then we have pretty similar logic in `AbstractStreamTaskNetworkInput#emitNext:121`

I find simply returning the event rather than an empty buffer a cleaner solution tbh.

Reply:
> but for EndOfPartitionEvent, CheckpointedInputGate returns an empty buffer and then we have pretty similar logic in

Does it? After looking at the code it looks to me as `EndOfPartitionEvent` is also returned from the input gate many times over and over again?

Still, I would prefer to have a consistency in handling those two events in the same way. So either  return here `MORE_AVAILABLE` always, or implement a similar check as this for `EndOfPartitionEvent`?

Reply:
Ok, you're right that all but last `EndOfPartitionEvent` are returned. There is a slight difference though. Once we received the last `EndOfPartitionEvent`, the `InputGate` will keep returning `Optional.empty()` from now on. Therefore we need the logic in 
`AbstractStreamTaskNetworkInput#emitNext:121`. We could copy that logic over to `processEvent` and do 
```
    protected DataInputStatus processEvent(BufferOrEvent bufferOrEvent) {
        // Event received
        final AbstractEvent event = bufferOrEvent.getEvent();
        if (event.getClass() == EndOfData.class) {
            // release the record deserializer immediately,
            // which is very valuable in case of bounded stream
            releaseDeserializer(bufferOrEvent.getChannelInfo());
            if (checkpointedInputGate.hasReceivedEndOfData()) {
                return DataInputStatus.END_OF_DATA;
            }
        } else if (event.getClass() == EndOfPartitionEvent.class) {
            if (checkpointedInputGate.isFinished()) {
                return DataInputStatus.END_OF_INPUT;
            }
        } else if (event.getClass() == EndOfChannelStateEvent.class) {
            if (checkpointedInputGate.allChannelsRecovered()) {
                return DataInputStatus.END_OF_RECOVERY;
            }
        }
        return DataInputStatus.MORE_AVAILABLE;
    }
```

This would I guess save us one invocation of `emitNext()` as I believe in the current master we must call the `emitNext` one more time after we received the last `EndOfPartitionEvent` to finally return `END_OF_INPUT` status. If we don't mind duplicating the check, I am fine adding this branch.

Reply:
> We could copy that logic over to processEvent and do

ok, let's unify it this way